### PR TITLE
Fix incorrect null handling for boolean kernels `and` and `or`

### LINKED ISSUE #2469: Incorrect null handling at boolean kernels
**Describe the bug**

null AND false should be false, instead of null. 
null OR true should be true, instead of null.

Currently `and` and `or` are incorrect on null handling.

**To Reproduce**

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #2469.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow/src/compute/kernels/boolean.rs [hunk 1]
```diff
@@ -174,14 +174,86 @@ where
...
+/// in boolean AND kernel. In short, because for AND kernel, null AND false results false.
+/// So we cannot simply AND two null buffers. This function updates null buffer of one side
+/// if other side is a false value.
+pub(crate) fn null_buffer_bin_and(
+    null_buffer: &Buffer,
+    null_buffer_offset_in_bits: usize,
+    data_buffer: &Buffer,
+    data_buffer_offset_in_bits: usize,
+    data_null_buffer: Option<&Buffer>,
+    len_in_bits: usize,
```
Comment (tustvold) [inline arrow/src/compute/kernels/boolean.rs:187]: I find the naming of these extremely confusing, in fact I'm still kind of confused what is going on...

File: arrow/src/compute/kernels/boolean.rs [hunk 2]
```diff
@@ -193,13 +265,56 @@ where
...
+        (Some(left_null_buffer), None) => Some(left_null_buffer),
+        (None, Some(right_null_buffer)) => Some(right_null_buffer),
+        (Some(left_null_buffer), Some(right_null_buffer)) => Some(buffer_bin_and(
+            &left_null_buffer,
+            0,
+            &right_null_buffer,
+            0,
+            len,
+        )),
+    };
```
Comment (tustvold) [inline arrow/src/compute/kernels/boolean.rs:316]: Why can't we do this in one go?

**Review by tustvold — COMMENTED** [review]
I'll take another look with fresh eyes tomorrow, I can't quite follow what is going on here :sweat_smile:

File: arrow/src/compute/kernels/boolean.rs (hunk 2 above)
Comment (viirya) [reply arrow/src/compute/kernels/boolean.rs:316]: I think it can be simplified in just one run. I will update it later.

Comment (viirya) [reply arrow/src/compute/kernels/boolean.rs:316]: It is the current form as I tried to code the solution. Not simplified yet.

Comment (viirya) [reply arrow/src/compute/kernels/boolean.rs]: I made building null buffer in a single run now. Also added some more comments. Hopefully it helps.

**Comment by tustvold** [comment]
Was this closed intentionally?

**Comment by viirya** [comment]
As @jhorstmann mentioned https://github.com/apache/arrow-rs/issues/2469#issuecomment-1217644505, `and_kleene` / `or_kleene` kernels was already introduced for AND and OR operations with SQL behavior. It seems to be intentional to keep this distinction (`and`/`and_kleene`, `or`/`or_kleene`). So I close this.

**Comment by viirya** [comment]
Another thought. `binary_boolean_kleene_kernel` looks complicated and maybe I can modify this change to replace current implementation and share code between `and`, `and_kleene`, `or` and `or_kleene` so they only use `binary_boolean_kernel`.

**Comment by viirya** [comment]
Another thought. `binary_boolean_kleene_kernel` looks complicated and maybe I can modify this change to replace current implementation and share code between `and`, `and_kleene`, `or` and `or_kleene` so they only use `binary_boolean_kernel`.

**Pull request closed without merging**